cts/cli/regression.tools.exp

```diff
@@ -3471,58 +3470,56 @@ Removing constraint: cli-prefer-dummy
 </pacemaker-result>
 =#=#=#= End test: Check locations and constraints for prim1 in XML - OK (0) =#=#=#=
 * Passed: crm_resource   - Check locations and constraints for prim1 in XML
 =#=#=#= Begin test: Recursively check locations and constraints for prim1 in XML =#=#=#=
 <pacemaker-result api-version="X" request="crm_resource -A -r prim1 --output-as=xml">
-  <constraints>
-    <resource id="prim1"/>
-  </constraints>
+  <constraints/>
   <status code="0" message="OK"/>
 </pacemaker-result>
 =#=#=#= End test: Recursively check locations and constraints for prim1 in XML - OK (0) =#=#=#=
 * Passed: crm_resource   - Recursively check locations and constraints for prim1 in XML
 =#=#=#= Begin test: Check locations and constraints for prim2 =#=#=#=
-prim2:
-  * Locations:
-    * Node cluster01 (score=INFINITY, id=prim2-on-cluster1)
-Colocations:
+Location preferences for prim2:
+  * Node cluster01 (score=INFINITY, id=prim2-on-cluster1)
+Resources prim2 is colocated with:
   * prim3 (score=INFINITY, id=colocation-prim2-prim3-INFINITY)
 =#=#=#= End test: Check locations and constraints for prim2 - OK (0) =#=#=#=
 * Passed: crm_resource   - Check locations and constraints for prim2
 =#=#=#= Begin test: Recursively check locations and constraints for prim2 =#=#=#=
-prim2:
-  * Locations:
-    * Node cluster01 (score=INFINITY, id=prim2-on-cluster1)
-Colocations:
+Location preferences for prim2:
+  * Node cluster01 (score=INFINITY, id=prim2-on-cluster1)
+Resources prim2 is colocated with:
   * prim3 (score=INFINITY, id=colocation-prim2-prim3-INFINITY)
+    * Resources prim3 is colocated with:
+      * prim4 (score=INFINITY, id=colocation-prim3-prim4-INFINITY)
+        * Location preferences for prim4:
+          * Node cluster02 (score=INFINITY, id=prim4-on-cluster2)
+        * Resources prim4 is colocated with:
+          * prim5 (score=INFINITY, id=colocation-prim4-prim5-INFINITY)
 =#=#=#= End test: Recursively check locations and constraints for prim2 - OK (0) =#=#=#=
 * Passed: crm_resource   - Recursively check locations and constraints for prim2
 =#=#=#= Begin test: Check locations and constraints for prim2 in XML =#=#=#=
 <pacemaker-result api-version="X" request="crm_resource -a -r prim2 --output-as=xml">
   <constraints>
-    <resource id="prim2">
-      <list name="locations" count="1">
+    <locations>
+      <location host="cluster01" id="prim2-on-cluster1" score="INFINITY"/>
+    </locations>
+    <rsc_colocation id="colocation-prim2-prim3-INFINITY" rsc="prim2" with-rsc="prim3" score="INFINITY" node-attribute="" rsc-role="Unknown" with-rsc-role="Unknown">
+      <locations>
         <location host="cluster01" id="prim2-on-cluster1" score="INFINITY"/>
-      </list>
-    </resource>
-    <colocations>
-      <rsc_colocation id="colocation-prim2-prim3-INFINITY" rsc="prim2" with-rsc="prim3" score="INFINITY" node-attribute="" rsc-role="Unknown" with-rsc-role="Unknown"/>
-    </colocations>
+      </locations>
+    </rsc_colocation>
```

Comment:
The CIB uses rsc_location (aka XML_CONS_TAG_RSC_LOCATION) for location constraints (and "node" instead of "host"), so we might as well continue with that. Also, we can go even flatter and ditch the locations/colocations elements, rsc_location and rsc_colocation will give enough info.

Reply:
Are you thinking something like this instead:
```
<constraints>
  <location node="cluster01" id="prim2-on-cluster1" score="INFINITY"/>
  <rsc_location id="colocation-prim2-prim3-INFINITY" rsc="prim2" with-rsc="prim3" score="INFINITY" node-attribute="" rsc-role="Unknown" with-rsc-role="Unknown">
    <location node="cluster01" id="prim2-on-cluster1" score="INFINITY"/>
  </rsc_location>
</constraints>
```

Reply:
Like:
```
<constraints>
  <rsc_location node="cluster01" id="prim2-on-cluster1" score="INFINITY"/>
  <rsc_colocation id="colocation-prim2-prim3-INFINITY" rsc="prim2" with-rsc="prim3" score="INFINITY" node-attribute="" rsc-role="Unknown" with-rsc-role="Unknown"/>
  <rsc_location node="cluster01" id="prim2-on-cluster1" score="INFINITY"/>
</constraints>
```

Reply:
Edited XML in previous comment. Looking at it again, I'm not sure why the location constraint is doubled in this PR's output. Maybe the recursive location message is being given the original resource rather than the peer. I also didn't notice that rsc was missing from the location constraints.

Reply:
I can fix all those in a followup commit in this PR.  The rsc has always been missing from location constraints because it wasn't necessary - we had that tree structure to the output, so you could figure out the resource from context.  Once I started messing around with it, that was no longer the case.  So yeah, I'll add it in.

Reply:
Location constraints are doubled only for the resource given on the command line with "-r <rsc>", I suppose because of the extra locations-list call in stacks_and_constraints.

Reply:
And appears to be fixed by replacing an errant cons->rsc_lh that should have been an rh.